conformance/utils/http/http.go

```diff
@@ -72,12 +79,49 @@ func MakeRequestAndExpectResponse(t *testing.T, r roundtripper.RoundTripper, gwA
 			req.Headers[name] = []string{value}
 		}
 	}
-	cReq, cRes, err := r.CaptureRoundTrip(req)
-	require.NoErrorf(t, err, "error making request")
 
+	cReq, cRes := WaitForConsistency(t, r, req, expected, numConsistencyChecksPerRequest)
 	ExpectResponse(t, cReq, cRes, expected)
 }
 
+// WaitForConsistency repeats the provided request until it completes with a response having
+// the expected status code consistently. The provided threshold determines how many times in
+// a row this must occur to be considered "consistent".
+func WaitForConsistency(t *testing.T, r roundtripper.RoundTripper, req roundtripper.Request, expected ExpectedResponse, threshold int) (*roundtripper.CapturedRequest, *roundtripper.CapturedResponse) {
+	var (
+		cReq         *roundtripper.CapturedRequest
+		cRes         *roundtripper.CapturedResponse
+		err          error
+		numSuccesses int
+	)
+
+	require.Eventually(t, func() bool {
+		cReq, cRes, err = r.CaptureRoundTrip(req)
+		if err != nil {
+			numSuccesses = 0
+			t.Logf("Request failed, not ready yet: %v", err.Error())
```

Comment:
Looking at the ways CaptureRoundTrip can return an error, I'm wondering if we should exit early with a test failure here like we were before (`require.NoErrorf(t, err, "error making request")`). For example, if the error we're getting is from `http.NewRequestWithContext` (https://github.com/kubernetes-sigs/gateway-api/blob/master/conformance/utils/roundtripper/roundtripper.go#L87), we're not going to have any more success retrying with the same params.

Reply:
`CaptureRoundTrip` will return an error in the event that Envoy has not stood up listeners on the gateway yet even if it's been "programmed", so I don't think we should fail the test at this point. The general progression we see in our conformance test runs is:

```log
# Envoy isn't ready yet, so this returns an error from CaptureRoundTrip
http.go:102: Request failed, not ready yet: Get "http://172.18.255.201/": dial tcp 172.18.255.201:80: connect: connection refused

http.go:114: Request has passed 1 times in a row of the desired 3, not ready yet
http.go:114: Request has passed 2 times in a row of the desired 3, not ready yet
http.go:118: Request has passed 3 times in a row of the desired 3, ready!
```
(taken from the **Run tests** step in our Actions workflow [here](https://github.com/hashicorp/consul-api-gateway/runs/5821189907?check_suite_focus=true))

Reply:
Good catch, that makes sense to me.